Data/ByteString/Lazy/Char8.hs

```diff
@@ -792,43 +797,122 @@ unwords :: [ByteString] -> ByteString
 unwords = intercalate (singleton ' ')
 {-# INLINE unwords #-}
 
--- | readInt reads an Int from the beginning of the ByteString.  If
--- there is no integer at the beginning of the string, it returns
--- Nothing, otherwise it just returns the int read, and the rest of the
--- string.
+-- | Bounds for Word# multiplication by 10 without overflow, and
+-- absolute values of Int bounds.
+intmaxWord, intminWord, intmaxQuot10, intmaxRem10, intminQuot10, intminRem10 :: Word
+intmaxWord = fromIntegral (maxBound :: Int)
+intminWord = fromIntegral (negate (minBound :: Int))
+(intmaxQuot10, intmaxRem10) = intmaxWord `quotRem` 10
+(intminQuot10, intminRem10) = intminWord `quotRem` 10
+
+-- | Try to read an 'Int' value from the 'ByteString', returning @Just (val,
+-- str)@ on success, where @val@ is the value read and @str@ is the rest of the
+-- input string.  If the sequence of digits decodes to a value larger than can
+-- be represented by an 'Int', the returned value will be 'Nothing'.
+--
+-- This function will not read an /unreasonably/ long stream of leading
+-- zero digits when trying to decode a number.  When reading the first
+-- non-zero digit would require requesting a new chunk and ~32KB of
+-- leading zeros have already been read, the conversion is aborted and
+-- 'Nothing' is returned.
```

Comment:
This basically means that the result of `readInt` depends on the chunk structure of the string. This is undesirable from my perspective, because chunks are implicit and should be indistinguishible from user perspective.

Reply:
But there's never any latent lazy IO in the middle of a chunk.  The distinction is only important under extremely unusual circumstances, 32k or more of leading zeros read so far, and the current chunk (more than one might have been consumed) has ended.  My take was that once the chunk is already in memory,  most of the cost has been paid, and we may as well keep reading.

Granted, with `streaming-bytestring` the chunk structure and monadic effects are more explicit, and I could perhaps even tweak the policy to ignore chunk boundaries, and only stop when a new monadic effect is required to fetch the next chunk.

Here the chunk structure is our only signal that more data may need to be brought into memory. Also, and perhaps more importantly, checking the zero count limit at a smaller granularity can have a noticeable performance impact. Chunk boundaries give us an opportunity to make the check infrequently.

Reply:
I do not suggest to check zero count limit at a smaller granularity. I'd rather do not check it at all. Otherwise a) the behaviour of `readInt` depends on the internal structure of chunks, b) lazy `readInt` behaves differently from strict `readInt`.

Reply:
I see, that's a tenable policy. Your take is that unlike the `streaming-bytestring` variant, here you'd prefer to just keep reading unconditionally.  If someone is unlucky enough to read an `Int` from an unbounded stream (network socket) of zeros, then they should have taken steps to limit the amount of data that might be consumed...

So it boils down to whether protection against unbounded reads is part of the "safety" in this new *safer* `readInt`. I can live with removing the limit, though for users who do want to avoid spinning forever the work-around is a bit messy.

One would have to use something like `readInt . take 32768`, but then under normal conditions with the integer taking just a few bytes of the input one is left with not the rest of the bytestring, but the rest of the first 32k, which is not particularly convenient.

If the bound is wanted, doing it around `readInt` is messy and inefficient, so I think your suggestion amounts to not offering the (lazy I/O) bound.  If that's acceptable (it is what you get with `readInteger` at the moment) then OK, if others agree that's what the API should do.

Reply:
So long as "lazy" bytestrings are not actually of the *lazy-io* variety, but are rather data already in memory, arranged in chunks with perhaps some deferred pure computations in the tail, one might argue that it is OK to not bound the `readInt` input buffer.

But even with pure lazy ByteStrings pathology is still possible, and avoided by this PR:
```
λ> import qualified Data.ByteString.Lazy.Char8 as L8
λ> let x = L8.repeat '0'
λ> L8.readInt x
Nothing
λ> L8.readInteger x
Just [... time goes by ...]^CInterrupted.
```

Reply:
Well, `length (L8.repeat '0')`  diverges as well, but nobody sets an upper bound for it. 

Yes, my take is that `streaming-bytestring` has more explicit notion of chunks and effects, does not have a strict counterpart and is not a core package, which allows to be more "opinionated" in design decisions. But in `bytestring` I would prefer lazy API be invariant modulo internal representation and mimic strict API as close as possible. @sjakobi @hsyl20 thoughts?

Reply:
No feedback seems likely, unless more prodding will help...  How shall we proceed?

Reply:
Sorry for the delay! I'll try to make up my mind until Sunday.

Reply:
Thanks for arguing your points so clearly, @Bodigrim and @vdukhovni, and apologies for the late response! :)

I agree with @Bodigrim's points:

> But in `bytestring` I would prefer lazy API be invariant modulo internal representation and mimic strict API as close as possible.